### PR TITLE
feat: add BPR implicit-feedback recommender

### LINKED ISSUE #16: feat: add implicit-feedback BPR recommender
Blocked: needs the common Recommender interface (and matrix/split utilities) to land first.

**Acceptance criteria**
- Bayesian Personalized Ranking for implicit feedback, on the interface, tested.

### PR DESCRIPTION
**Pull request by Burton-David**
Bayesian Personalized Ranking on the unified Recommender interface, in ~80 lines of pure numpy. No PyTorch, no heavy dep — the SGD loop is straightforward.

## How it works

Every observed (user, item) pair is a positive signal; a random unobserved item per positive is sampled as a negative. The objective is to score positives higher than their sampled negatives, optimized via SGD on the sigmoid-margin loss with L2 regularization. Rating *values* are ignored — only the presence of an interaction matters, which is what makes this an implicit-feedback model.

The update step for each (u, i, j) triple:

```
margin = U[u] · (V[i] - V[j])
σ      = 1 / (1 + exp(margin))         # sigmoid(-margin)
U[u]  += lr * (σ · (V[i] - V[j]) - reg · U[u])
V[i]  += lr * (σ · U[u]            - reg · V[i])
V[j]  += lr * (-σ · U[u]           - reg · V[j])
```

Subclasses `_MatrixBackedRecommender` (from #35), so `BPR` only needs to fit and produce per-user scores — the unknown-user fast path, seen-mask, and top-n filter come for free.

## Tests (5 new, 81 total)

- Interface conformance.
- **Cluster recovery on synthetic data**: 50 users in two clusters, each interacting with 6 items from their cluster. After 40 epochs the top-10 recommendations for a cluster-0 user land ≥8/10 in cluster-0 items — strong evidence the embeddings actually separate the structure.
- Exclude-seen is enforced.
- Unknown-user fast path returns `[]`.
- Seeded fits are reproducible.

## Local checks

```
ruff check src tests          # clean
ruff format --check src tests # clean
mypy                          # Success: no issues found in 12 source files
pytest                        # 81 passed
```

Closes #16

### CODE REVIEW

**Review by JohnJacob-coder — CHANGES_REQUESTED** [review]
The BPR math is correct — I checked the gradient: `sig = 1/(1+exp(margin)) = σ(−margin)`, and the three updates (`u += lr·(sig·(i−j) − reg·u)`, `i += lr·(sig·u − reg·i)`, `j += lr·(−sig·u − reg·j)`) are exactly the BPR-OPT ascent steps. And `test_bpr_separates_clusters` is a real learning test, not a smoke test. Nice work.

One blocker and two suggestions:

**Blocking — possible infinite loop in negative sampling (`bpr.py:73`):**
```python
while observed[u, j]:
    j = int(rng.integers(0, n_items))
```
If user `u` has interacted with *every* item (all of `observed[u]` is `True`), this never terminates — it hangs the process with no error. Sparse real data won't hit it, but a small/degenerate dataset or a "rated everything" user will. The tests miss it only because each synthetic user touches 6 of 50 items. Please guard it — e.g. skip users with no unobserved items, or cap resample attempts and `continue`. A test with one all-items user would lock the guard in.

**Suggestion (non-blocking) — the overflow comment:** `1/(1+exp(margin))` still overflows `exp` for large *positive* margin (emits a RuntimeWarning; the result correctly → 0). Either clip the margin / use a stable form, or soften the comment from "avoids overflow" to "saturates safely."

**Suggestion (non-blocking) — performance:** the per-sample Python loop is O(epochs × positives) with per-step indexing — ~1.6M iterations on full MovieLens-100k, so it'll be slow in the benchmark. Fine for a readable reference implementation, but worth a note or a later vectorization pass.

Typing, structure, and the rest of the tests are solid.

**Comment by Burton-David** [comment]
Thanks for catching the infinite-loop case — that was a real bug that the cluster-recovery test couldn't see.

**Blocker addressed.** Pre-filtered the training set so SGD only iterates over interactions from users that have at least one unobserved item:

```python
has_negatives = observed.sum(axis=1) < n_items
positives = np.argwhere(observed & has_negatives[:, None])
```

The inner `while observed[u, j]` is now guaranteed to terminate within at most `n_items` iterations.

Added `test_bpr_terminates_when_a_user_has_rated_every_item` which builds exactly the pathological case you described — one user who has rated every item alongside well-behaved users — and asserts that fit completes (no hang) and that user's recommendations are empty (everything filtered as seen).

**Suggestion #1 (overflow comment).** Softened to "saturates safely at the tails." A proper clip pass would be a worthwhile follow-up if we ever benchmark on real data with extreme margins.

**Suggestion #2 (vectorization).** Agreed it'll be slow on MovieLens-100k — happy to file a follow-up issue for a vectorized BPR pass that batches the per-step updates. Out of scope for this PR though; the readable reference impl is the right starting point.

Local: 82 tests pass, including the new termination guard; mypy/ruff clean. Pushed `6653412`.

**Review by JohnJacob-coder — APPROVED** [review]
Guard is correct — restricting positives to users with at least one unobserved item guarantees the resample loop terminates, and the all-items-user regression test locks it in. Math was already right; this closes the hang. The two non-blocking notes (overflow comment, per-sample loop perf) are fine to leave as future polish. Approving.
